src/udiskslinuxblock.c

```diff
@@ -1175,8 +1181,25 @@ udisks_linux_block_update (UDisksLinuxBlock       *block,
       udisks_block_set_id (iface, NULL);
     }
 
-  udisks_block_set_id_usage (iface, g_udev_device_get_property (device->udev_device, "ID_FS_USAGE"));
-  udisks_block_set_id_type (iface, g_udev_device_get_property (device->udev_device, "ID_FS_TYPE"));
+  seems_encrypted = bd_crypto_device_seems_encrypted (device_file, &error);
```

Comment:
If I understand this correctly, this will run the check on every device. I really don't like this -- we should at least run the check only on devices that "look empty" ("ID_FS_USAGE" udev property is empty).

I generally don't like the idea of running this automatically on devices -- we generally try to not read "random" devices (we take most of our information from udev) and it is possible this will fail for some "special" devices. And it can also produce some "false positives" -- for example LUKS with detached header will be detected as encrypted (it is encrypted, but we can't open it, because we don't support opening LUKS with detached header). This can be really confusing for gvfs/Nautilus and other "users".
I think we should consider doing some "whitelisting" for devices we want to check -- e.g. if user wants to run this check on a device, he has to create a special udev rule adding a property that will tell udisks not check the devices (something similar to `UDISKS_IGNORE` we currently use to ignore "system" devices). We unfortunately don't have a config file for udisks, so udev rules are only possibility for something like this, but it is quite easy to create a new one and it is even possible to create an udev rule "outside" of udisks -- e.g. if some distribution decides they want to switch this on for all devices, it is easy to create such rule without patching udisks.

Any ideas/thoughts about this? I'm open to ideas/discussion, it is possible I'm being too scared without a reason, or my solution is overly complicated.

Reply:
> If I understand this correctly, this will run the check on every device. 

Correct.

> I really don't like this -- we should at least run the check only on devices that "look empty" ("ID_FS_USAGE" udev property is empty).

I would be okay with that, but it is something we had already considered and decided against, so let me explain why: The `ID_FS_USAGE` property is set based on the filesystem magic number, which is usually only 2 bytes. So if a device which contains random data has the wrong bytes at a certain offset, it will be incorrectly identified as filesystem-formatted. As a result, udisks will not be able to unlock a given `TCRYPT` volume with a chance of `n` / 65536, where `n` = number of filesystems with a 2 byte magic number that are supported by udev.

> I generally don't like the idea of running this automatically on devices -- we generally try to not read "random" devices (we take most of our information from udev) 

I understand that, and I had the same feeling when I first realized that this will be required to detect TCRYPT devices. But we only try to read the first 512 bytes of the device, which udev does anyway, in order to parse the superblock. So I don't see why it would cause problems if we do that in libbockdev / udisks.

> and it is possible this will fail for some "special" devices.

But then `bd_crypto_device_seems_encrypted` would simply return `FALSE`, no?  

> And it can also produce some "false positives" -- for example LUKS with detached header will be detected as encrypted (it is encrypted, but we can't open it, because we don't support opening LUKS with detached header). This can be really confusing for gvfs/Nautilus and other "users".

Absolutely, all LUKS volumes with a detached header will pass this test, as well as dm-crypt plain and Loop-AES volumes. To make this less confusing, we introduced the `crypto_unknown` type, and in our GNOME Disks patches we add a sentence to the unlock dialog explaining that "this volume might be a VeraCrypt volume, as it contains random data".

> I think we should consider doing some "whitelisting" for devices we want to check -- e.g. if user wants to run this check on a device, he has to create a special udev rule adding a property that will tell udisks not check the devices (something similar to UDISKS_IGNORE we currently use to ignore "system" devices). We unfortunately don't have a config file for udisks, so udev rules are only possibility for something like this, but it is quite easy to create a new one and it is even possible to create an udev rule "outside" of udisks -- e.g. if some distribution decides they want to switch this on for all devices, it is easy to create such rule without patching udisks.

I think this feature will be pretty useless if it requires the user to create a udev rule first, because if they are able to do that, they are also able to use `cryptsetup` on the command line to unlock their volumes.
I also don't see how this can work for encrypted file containers, which can be unlocked by associating them with a loop device.

Reply:
> I would be okay with that, but it is something we had already considered and decided against, so let me explain why: The ID_FS_USAGE property is set based on the filesystem magic number, which is usually only 2 bytes. So if a device which contains random data has the wrong bytes at a certain offset, it will be incorrectly identified as filesystem-formatted. As a result, udisks will not be able to unlock a given TCRYPT volume with a chance of n / 65536, where n = number of filesystems with a 2 byte magic number that are supported by udev.

That's a good point. On the other hand if udev thinks something is a filesystem udisks won't be the only one that is really confused. It might be worth testing what happens in that case (maybe nothing).

> I understand that, and I had the same feeling when I first realized that this will be required to detect TCRYPT devices. But we only try to read the first 512 bytes of the device, which udev does anyway, in order to parse the superblock. So I don't see why it would cause problems if we do that in libbockdev / udisks.

Udev is caching a lot of information. With this change udisks will do this check for every udev change event on every block device. We already have some issues about udisks starting too long or using too much CPU because of some bugged devices.

> I think this feature will be pretty useless if it requires the user to create a udev rule first, because if they are able to do that, they are also able to use cryptsetup on the command line to unlock their volumes.

You'd need to do that only once and than use udisks for unlocking, but is probably too complicated. Too bad we don't have some better/easier way of configuration.

Maybe I'm just too paranoid or scared of breaking something. Any thoughts @vpodzime ?

Reply:
> Maybe I'm just too paranoid or scared of breaking something. Any thoughts @vpodzime ?

What about simply adding some "flag file" like ``/etc/udev/tcrypt.conf`` that would enable all this machinery? It could even be empty (for now) and just tell *udisks* TrueCrypt/VeraCrypt devices should be taken into account. It can then be packaged separately and people who are using TC/VC would simply install this extra package.

Reply:
Opinions @segfault3, @vojtechtrefny?

Reply:
Sorry, I was waiting for opinions of my team mates (which I didn't get yet). Anyway, I think this should work for us (Tails), because we can simply ship this file. But it doesn't seem super intuitive to me to have a config file without any options in it. Maybe it would be better to create `/etc/udisks2.conf` and set an option like `TcryptEnable=True`?

Reply:
> This would not provide any easy/nice way to make the support installable as a package. Packagers would have to decide what the default should be and then people would have to change it manually.

How would that be different when using the flag file? The way I understand it, packagers would still have to decide what the default should be (ship the flag file or not), and then people would have to change it manually (create the flag file or delete it).

But I don't have a strong opinion on this, if you think the flag file is better, then it's fine for me - it's definitely easier to implement.

Reply:
The flag file could be put into a separate package like *udisks2-truecrypt*. And at some point the file could contain some configuration for the TC/VC-related behaviour.

Reply:
Right. Ok, I will implement the check for the flag file then.

Reply:
> Right. Ok, I will implement the check for the flag file then.

Thanks! Later we can implement some real configuration like blacklist and/or whitelist of devices that should be scanned, but for now this is an easy way to prevent potential confusing issues for users that don't use TC/VC.

Reply:
Sorry for the delay. I pushed a commit now (03781dd).